File: go/pkg/2024/day3.go
package solver2024

import (
	"regexp"
	"strconv"
)

type SolverThree struct {
}

func (solver SolverThree) SolutionOne(lines []string) int {
    answer := 0

    for _, line := range lines {
        if line == "" {
            continue
        }

        multiplications := findMultiplications(line)

        for _, multiplication := range multiplications {
            answer += multiplyElement(multiplication)
        }
    }
    return answer
}

func (solver SolverThree) SolutionTwo(lines []string) int {
    answer := 0
    enabled := true

    for _, line := range lines {
        if line == "" {
            continue
        }

        instructions := findExtendedInstructions(line)

        for _, instruction := range instructions {
            switch instruction {
            case "do()":
                enabled = true
            case "don't()":
                enabled = false
            default:
                if enabled {
                    answer += multiplyElement(instruction)
                } else {
                    continue
                }
            }
        }
    }
    return answer
}

func findMultiplications(line string) []string {
    regex, err := regexp.Compile(`mul\([0-9]{1,3},[0-9]{1,3}\)`)
    if err != nil {
        panic("Invalid Regex")
    }

    return regex.FindAllString(line, -1)
}

func findExtendedInstructions(line string) []string {
    regex, err := regexp.Compile(`(mul\([0-9]{1,3},[0-9]{1,3}\))|(do\(\))|(don't\(\))`)
    if err != nil {
        panic("Invalid Regex")
    }

    return regex.FindAllString(line, -1)
}

func multiplyElement(element string) int {
    regex, err := regexp.Compile(`[0-9]{1,3}`)

    if err != nil {
        panic("Invalid Regex 2")
    }

    numbers := regex.FindAllString(element, -1)

    if len(numbers) != 2 {
        panic("2 numbers not found")
    }

    num1, _ := strconv.Atoi(numbers[0])
    num2, _ := strconv.Atoi(numbers[1])

    return num1 * num2
}

File: rust/aoc/src/solver/s24/day3.rs
use crate::solver::solver::Solver;

pub struct Day3 {
}

impl Solver for Day3 {
    fn solution_one(&self, lines: Vec<String>) -> i64 {
        let mut answer: i64 = 0;

        for line in lines {
            let multipliers = get_multipliers(line);

            for multiplier in multipliers {
                answer += multiply(multiplier);
            }
        }

        return answer;
    }

    fn solution_two(&self, lines: Vec<String>) -> i64 {
        let mut answer: i64 = 0;
        let mut enable: bool = true;

        for line in lines {
            let instructions = get_all_instructions(line);

            for instruction in instructions {
                if instruction == "do()" {
                    enable = true;
                    continue;
                } else if instruction == "don't()" {
                    enable = false;
                    continue;
                } else if enable {
                    answer += multiply(instruction);
                }
            }
        }
        
        return answer;
    }
}

fn get_multipliers(line: String) -> Vec<String> {
    let regx = regex::Regex::new(r"mul\(\d{1,3},\d{1,3}\)").unwrap();

    return regx.find_iter(&line).map(|e| String::from(e.as_str())).collect();
}

fn get_all_instructions(line: String) -> Vec<String> {
    let regx = regex::Regex::new(r"do\(\)|mul\(\d{1,3},\d{1,3}\)|don't\(\)").unwrap();

    return regx.find_iter(&line).map(|e| String::from(e.as_str())).collect();
}

fn multiply(multiplier: String) -> i64 {
    dbg!(&multiplier);
    let regx = regex::Regex::new(r"\d+").unwrap();

    let numbers: Vec<i64> = regx.find_iter(&multiplier).map(|e| e.as_str().parse::<i64>().unwrap()).collect();

    if numbers.len() != 2 {
        return 0;
    }

    return numbers[0] * numbers[1];
}
